dll/win32/shell32/wine/control.c

```diff
@@ -967,65 +960,199 @@ static	void	Control_DoLaunch(CPanel* panel, HWND hWnd, LPCWSTR wszCmd)
+        else
+        {
+            lstrcpyW(buffer, wszCmd);
+        }
+    }
+    /* If an unquoted comma was found, there are at least two parts of the string:
+     * - the CPL path
+     * - either a dialog box number preceeded by @, or a dialog box name.
+     * If there was a second unqoted comma, there is another part of the string:
+     * - the rest of the parameters. */
+    else
+    {
+        /* If there was no second unquoted comma in the string, the CPL path ends at thes
+          * null terminator. */
+        if (wszSecondCommaPosition == NULL)
+            wszSecondCommaPosition = wszCmd + nLen;
+
+        StringCchCopyNW(buffer, nLen, wszCmd, wszFirstCommaPosition - wszCmd);
+        StringCchCopyNW(wszDialogBoxName,
+                        nLen,
+                        wszFirstCommaPosition + 1,
+                        wszSecondCommaPosition - wszFirstCommaPosition - 1);
+
+        if (wszSecondCommaPosition != wszCmd + nLen)
+        {
+            extraPmts = wszSecondCommaPosition + 1;
+        }
+    }
 
-        aCPLFlags = GlobalFindAtomW(L"CPLFlags");
-        if (!aCPLFlags)
+    /* Remove the quotes from both buffers. */
+    while ((ptr = StrChrW(buffer, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
+
+    while ((ptr = StrChrW(wszDialogBoxName, '"')))
+        memmove(ptr, ptr+1, lstrlenW(ptr)*sizeof(WCHAR));
+
+    if (wszDialogBoxName[0] == L'@')
+    {
+        sp = atoiW(wszDialogBoxName + 1);
+    }
+
+    TRACE("cmd %s, extra %s, sp %d\n", debugstr_w(buffer), debugstr_w(extraPmts), sp);
+
+    applet = Control_LoadApplet(hWnd, buffer, panel);
+    if (applet)
+    {
+        ULONG_PTR cookie;
+        BOOL bActivated;
+        ATOM aCPLName;
+        ATOM aCPLFlags;
+        ATOM aCPLPath;
+        AppDlgFindData findData;
+        
+        /* we've been given a textual parameter (or none at all) */
+        if (sp == -1) 
         {
-            aCPLFlags = GlobalAddAtomW(L"CPLFlags");
+            while ((++sp) != applet->count) 
+            {
+                TRACE("sp %d, name %s\n", sp, debugstr_w(applet->info[sp].name));
+
+                if (StrCmpIW(wszDialogBoxName, applet->info[sp].name) == 0)
+                    break;
+            }
         }
 
-        findData.szAppFile = applet->cmd;
-        findData.sAppletNo = (UINT_PTR)(sp + 1);
-        findData.aCPLName = aCPLName;
-        findData.aCPLFlags = aCPLFlags;
-        findData.hRunDLL = applet->hWnd;
-        findData.hDlgResult = NULL;
-        // Find the dialog of this applet in the first instance.
-        // Note: The simpler functions "FindWindow" or "FindWindowEx" does not find this type of dialogs.
-        EnumWindows(Control_EnumWinProc, (LPARAM)&findData);
-        if (findData.hDlgResult)
+        if (sp >= applet->count && wszDialogBoxName[0] == L'\0')
         {
-            BringWindowToTop(findData.hDlgResult);
+            sp = 0;
```

Comment:
The original WARN Wine had doesn't apply?
`WARN("Out of bounds (%u >= %u), setting to 0\n", sp, applet->count);`

Reply:
It does not - in wine it was a weird fallback option to use the index 0 dialog for when the dialog wasn't found.  Here it handles a narrower case - dialog name being empty. It is handled like that on purpose, so I don't see any reason to WARN.